pants-plugins/internal_plugins/versioned_setup/register.py

```diff
@@ -0,0 +1,81 @@
+import os.path
+
+from pants.backend.python.goals.setup_py import SetupKwargs, SetupKwargsRequest
+from pants.engine.fs import DigestContents, GlobMatchErrorBehavior, PathGlobs
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import Target
+from pants.engine.unions import UnionRule
+
+
+class VersionedSetupKwargsRequest(SetupKwargsRequest):
+    @classmethod
+    def is_applicable(cls, _: Target) -> bool:
+        # We always use our custom `setup()` kwargs generator for `python_distribution`
+        # targets in this repo.
+        return True
+
+
+@rule
+async def setup_kwargs_plugin(request: VersionedSetupKwargsRequest) -> SetupKwargs:
+    kwargs = request.explicit_kwargs.copy()
+
+    # Add classifiers. We preserve any that were already set.
+    standard_classifiers = [
+        "Intended Audience :: Developers",
+        "License :: OSI Approved :: MIT",
+        "Operating System :: MacOS :: MacOS X",
+        "Operating System :: POSIX :: Linux",
+        "Programming Language :: Python",
+        "Topic :: Software Development :: Build Tools",
+    ]
+    kwargs["classifiers"] = [*standard_classifiers, *kwargs.get("classifiers", [])]
+    project_urls = {
+        "Source": "https://github.com/CodeForAfrica/api",
+        "Tracker": "https://github.com/CodeForAfrica/api/issues",
+        "Twitter": "https://twitter.com/Code4Africa",
+    }
+    kwargs["project_urls"] = {**project_urls, **kwargs.get("project_urls", {})}
+    default_kwargs = dict(
+        url="https://github.com/CodeForAfrica/api",
+        author="Code for Africa",
+        author_email="tech@codeforafrica.org",
+        license="MIT",
+        zip_safe=True,
+    )
+    for key, value in default_kwargs.items():
+        if key not in kwargs:
+            kwargs[key] = value
```

Comment:
Haven't tried it but we could have something like `kwargs = default_kwargs | kwargs` right @kilemensi ? 
This way `kwargs` values are always used incase they appear in  `default_kwargs` too.

Reply:
Are you talking about about [`union`](https://peps.python.org/pep-0584/) @esirK? If yes, then yeah, I think all of your suggestions would work. I just assumed `kwargs` to be much bigger than `default_kwargs` and hence didn't want to create a copy of `kwargs` if we don't have to.

Reply:
yeah that's the suggestion @kilemensi

Reply:
Cool switched to `|=` @esirK

Reply:
NVM I think it's better this way 🚀